File: javascript/tabulation/bestSum.js
const bestSum = (targetSum, numbers) => {
  const table = Array(targetSum + 1).fill(null);
  table[0] = [];

  for (let i = 0; i <= targetSum; i++) {
    if (table[i] !== null) {
      for (let numb of numbers) {
        const combination = [...table[i], numb];
        if (!table[i+numb] || combination.length < table[i + numb].length) {
          table[i + numb] = combination;
        }
      }
    }
  }
  return table[targetSum];
};

console.log(bestSum(12, [3, 4, 5, 6]));

File: ruby/tabulation/bestSum.rb
def best_sum(targetSum, numbers)
    table = Array.new(targetSum+1, nil)
    table[0] = []

    (targetSum+1).times do |i|
        if table[i] != nil
            numbers.each do |n|
                combination = [*table[i], n]
                if (!table[i+n] || table[i+n].length > combination.length)
                    table[i+n] = combination
                end

            end
        end
    end

    table[targetSum]
end

p best_sum(12, [3, 4,6])